File: lessons/compiled_lang/rust/src/f02.loop.rs
// testbox: title="manual iterator (while let Some(...) = iter.next())"
use std::fs;

fn main() {
    let mut entries: Vec<_> = fs::read_dir("dirtest").unwrap().map(|e| e.unwrap()).collect();
    entries.sort_by_key(|e| e.file_name());

    // step through the iterator protocol a for-in loop desugars to
    let mut iter = entries.iter();
    while let Some(entry) = iter.next() {
        let name = entry.file_name().to_string_lossy().to_string();
        if entry.path().is_dir() {
            println!("{} is a directory", name);
        } else {
            println!("{} is not a directory", name);
        }
    }
}

File: lessons/compiled_lang/cpp/src/f02.loop.cpp
// testbox: title="manual iterator (begin()/end(), ++it, *it)"
#include <algorithm>
#include <filesystem>
#include <iostream>
#include <string>
#include <vector>

namespace fs = std::filesystem;

int main() {
    std::vector<std::string> items;
    for (const auto& entry : fs::directory_iterator("dirtest"))
        items.push_back(entry.path().filename().string());
    std::sort(items.begin(), items.end());

    // manual iterator loop - what a range-based for desugars to
    for (auto it = items.begin(); it != items.end(); ++it) {
        const std::string& item = *it;
        if (fs::is_directory("dirtest/" + item))
            std::cout << item << " is a directory" << std::endl;
        else
            std::cout << item << " is not a directory" << std::endl;
    }

    return 0;
}
